File: Week5/Minimum-Path-Sum/xuwei.cpp
class Solution {
public:
    int minPathSum(vector<vector<int> > &grid) {
        int m = grid.size();
        if(m == 0) 
            return 0;
        int n = grid[0].size();
        int **opt = new int*[m];
        for(int i=0;i<m;i++)
        {
            opt[i] = new int[n];
        }
        opt[0][0] = grid[0][0];
        for(int i = 1; i < m; i++)
        {
            opt[i][0] = opt[i-1][0] + grid[i][0];
        }
        for(int j = 1; j < n; j++)
        {
            opt[0][j] = opt[0][j-1] + grid[0][j];
        }
        for(int i = 1; i < m; i++)
        {
            for(int j = 1; j < n; j++)
            {
                opt[i][j] = min(opt[i][j-1],opt[i-1][j])+grid[i][j];
            }
        }
        int ret = opt[m-1][n-1];
        for(int i=0;i<m;i++)
        {
            delete [] opt[i];
        }
        delete[] opt;
        return ret;
    }
};

File: Week5/Minimum-Path-Sum/xuwei.py
class Solution:
    def minPathSum(self, grid):
        if len(grid) == 0:
            return 0
        dp = grid
        m,n = len(grid),len(grid[0])
        for i in range(1,m):
            dp[i][0] = dp[i-1][0] + grid[i][0]
        for i in range(1,n):
            dp[0][i] = dp[0][i-1] + grid[0][i]
        for i in range(1,m):
            for j in range(1,n):
                dp[i][j] = min(dp[i][j-1],dp[i-1][j]) + grid[i][j]
        return dp[m-1][n-1]
